output/cloud/expv2/hdr.go

```diff
@@ -0,0 +1,270 @@
+// We don't expect to have this case but the current API
+// would support it so we need to make sure it works or refactor.
+func (h *histogram) addToBucket(v float64) {
+	if h.Count == 0 {
+		h.Max, h.Min = v, v
+	} else {
+		if v > h.Max {
+			h.Max = v
+		}
+		if v < h.Min {
+			h.Min = v
+		}
+	}
+
+	h.Count++
+	h.Sum += v
+
+	if v > highestTrackable {
+		h.ExtraHighBucket++
+		return
+	}
+	if v < lowestTrackable {
+		h.ExtraLowBucket++
+		return
+	}
+
+	index := resolveBucketIndex(v)
+	blen := uint32(len(h.Buckets))
+	if blen == 0 {
+		h.FirstNotZeroBucket = index
+		h.LastNotZeroBucket = index
+	} else {
+		if index < h.FirstNotZeroBucket {
+			h.FirstNotZeroBucket = index
+		}
+		if index > h.LastNotZeroBucket {
+			h.LastNotZeroBucket = index
+		}
+	}
+
+	if index >= blen {
+		h.grow(index)
+	}
+	h.Buckets[index]++
+}
+
+// grow expands the buckets slice
+// with zeros up to the required index
+func (h *histogram) grow(index uint32) {
+	i := int(index)
+	if len(h.Buckets)-1 > i {
+		panic("buckets is already bigger than requested index")
+	}
+	if cap(h.Buckets) > i {
+		// See https://go.dev/ref/spec#Slice_expressions
+		// "For slices, the upper index bound is
+		// the slice capacity cap(a) rather than the length"
+		h.Buckets = h.Buckets[:index+1]
+	} else {
+		length := i + 1
+		// let's make two times larger of
+		// the current request
+		newBuckets := make([]uint32, length, (len(h.Buckets)+length)*2)
+		copy(newBuckets, h.Buckets)
+		h.Buckets = newBuckets
+	}
+}
+
+// trimzeros removes all buckets that have a zero value
+// from the begin and from the end until
+// the first not zero bucket.
+func (h *histogram) trimzeros() {
+	if h.Count < 1 || len(h.Buckets) < 1 {
+		return
+	}
+
+	// all the counters are set to zero, we can remove all
+	if h.FirstNotZeroBucket == 0 && h.LastNotZeroBucket == 0 {
+		h.Buckets = []uint32{}
+		return
+	}
+
+	h.Buckets = h.Buckets[h.FirstNotZeroBucket : h.LastNotZeroBucket+1]
+}
+
+// histogramAsProto converts the histogram into the equivalent Protobuf version.
+func histogramAsProto(h *histogram, time time.Time) *pbcloud.TrendHdrValue {
+	hval := &pbcloud.TrendHdrValue{
+		Time:              timestamppb.New(time),
+		MinResolution:     1.0,
+		SignificantDigits: 2,
+		LowerCounterIndex: h.FirstNotZeroBucket,
+		MinValue:          h.Min,
+		MaxValue:          h.Max,
+		Sum:               h.Sum,
+		Count:             h.Count,
+		Counters:          h.Buckets,
+	}
+	if h.ExtraLowBucket > 0 {
+		hval.ExtraLowValuesCounter = &h.ExtraLowBucket
+	}
+	if h.ExtraHighBucket > 0 {
+		hval.ExtraHighValuesCounter = &h.ExtraHighBucket
+	}
+	return hval
+}
+
+// resolveBucketIndex returns the index
+// of the bucket in the histogram for the provided value.
+func resolveBucketIndex(val float64) uint32 {
+	// the lowest trackable value is zero
+	// negative number are not expected
+	if val < 0 {
+		return 0
+	}
+
+	upscaled := uint32(math.Ceil(val))
+
+	// k is a power of 2 closest to 10^precision_points
```

Comment:
```suggestion
    // In histograms, bucket boundaries are usually defined as multiples of powers of 2, 
 	// allowing for efficient computation of bucket indexes.
 	//
 	// We define k=7 in our case, because it allows for sufficient granularity in the
 	// distribution (2^7=128 primary buckets of which each can be further 
 	// subdivided if needed). 
 	//
 	// k is the constant balancing factor between granularity and 
 	// computational efficiency.
 	//
 	// In our case:
 	// i.e 2^7  = 128  ~  100 = 10^2
 	//     2^10 = 1024 ~ 1000 = 10^3
 	// f(x) = 3*x + 1 - empiric formula that works for us
 	// since f(2)=7 and f(3)=10
```

Reply:
Done